graphql-protocol/src/main/java/com/linecorp/armeria/server/graphql/protocol/AbstractGraphqlService.java

```diff
@@ -63,8 +66,20 @@ public abstract class AbstractGraphqlService extends AbstractHttpService {
     private static final TypeReference<Map<String, List<String>>> MAP_PARAM =
             new TypeReference<Map<String, List<String>>>() {};
 
+    @Override
+    protected HttpResponse doConnect(ServiceRequestContext ctx, HttpRequest req) throws Exception {
+        if (ctx.sessionProtocol().isExplicitHttp2() && isHttp2WebSocketUpgradeRequest(req.headers())) {
```

Comment:
Sorry for the late review. 

There are two issues with the current approach.

1. `DefaultWebSocketService` is exposed as a public API. I understood why it was necessary. There was no extension point on `WebSocketService` API design.

2. The WebSocket handshake is a bit hard-coded in `DefaultGraphqlService`. We may delegate the handshake to `WebSocketService` and implement two methods. One is for WebSocket and another is for normal HTTP requests.

Proposal: 

Make `WebSocketService` abstract class so that `GraphqlService` can support WebSocket by extending or composing it.

```java
abstract class WebSocketService {

   protected abstract WebSocket handleWebSocket(ServiceRequestContext ctx, WebSocket in);

   protected HttpResponse handleHttp(ServiceRequestContext ctx, HttpRequest req) {
        return HttpResponse.of(HttpStatus.BAD_REQUEST, MediaType.PLAIN_TEXT_UTF_8,
                               "The upgrade header must contain");
   }
}
```
By changing the design of `WebSocketService`
1) We may split WebSocket logic from GraphQL over HTTP.
```java
class GraphqlWebSocketService extends WebSocketService implements GraphqlService {

  private final GraphqlService GraphqlService graphqlService;
   
  @Override
  protected WebSocket handleWebSocket(ServiceRequestContext ctx, WebSocket in) {
    // execute Graphql over WebSocket
  }

  @Override
  protected HttpResponse handleHttp(ServiceRequestContext ctx, HttpRequest req) {
    // delegate to the original GraphqlService
    return graphqlService.serve(ctx, req);
  }
}
```
2) Two GraphQL services can only be composed when WebSocket is enabled.
```java
class GraphqlServiceBuilder {

  GraphqlService build() {
    DefaultGraphqlService graphqlService = ...;
    if (enableWebSocket) {
       return new GraphqlWebSocketService(graphqlService);
    } else {
      return graphqlService;
    }
  }
}
```

If this proposal is acceptable, let me send a PR to refactor `WebSocketService` first.

Reply:
> DefaultWebSocketService is exposed as a public API. I understood why it was necessary. There was no extension point on WebSocketService API design.

I don't think this is necessarily due to a design choice. For instance, I think it is trivial to modify the current `DefaultWebSocketService` to be package-private in this pull request

> The WebSocket handshake is a bit hard-coded in DefaultGraphqlService. We may delegate the handshake to WebSocketService and implement two methods. One is for WebSocket and another is for normal HTTP requests.

I agree with this sentiment

Reply:
After discussion with other maintainers, I updated the API design to easily compose `WebSocketService`. #5368

Reply:
Nice. So with #5368 merged, if I understand it correctly, it would be possible to implement this entire PR by any user of the API without modifying Armeria itself?

Making this PR technically unnecessary, but just still a nice-to-have?

Reply:
> it would be possible to implement this entire PR by any user of the API without modifying Armeria itself?

Right. The new API in `WebSocketService` also allows to set a fallback `HttpService` for non-WebSocket requests. That means we can serve `WebSocketService` and `HttpService` at the same path.

> Making this PR technically unnecessary, but just still a nice-to-have?

We don't need to check WebSocket upgrade in `DefaultGraphqlService directly` but I believe `GraphqlWebSocketSubscriber` and related classes are still worth adding.

If #5368 is merged, GraphQL over WebSocket will be supported by the following steps:
```java
class GraphqlServiceBuilder {
   GraphqlService build() {
    // DefaultGraphqlService is used as is without modifying
    var graphqlService = ...
    if (webSocketEnabled) {
      val webSocketService = 
      WebSocketService
        .builder(graphqlWebSocketHandler)
        .fallbackService(graphqlService)
        ...
        .build()
      graphqlService = new GraphqlWebSocketService(webSocketService)
    }
    return graphqlService;
  }
}
```

`GraphqlWebSocketService` will be a simple wrapper that implements both `GraphqlService` and `WebSocketService`.
```java
class GraphqlWebSocketService implements GraphqlService, WebSocketService {
   WebSocketService delegate;

   @Override
   public WebSocket serve(ServiceRequestContext ctx, WebSocket in) {
     return delegate.serve(ctx, in);
   }
   ...
}
```

Reply:
> class GraphqlWebSocketService implements GraphqlService

Note: I prefer that `GraphqlWebSocketService` does not inherit `WebSocketService` so that HTTP requests do not unintentionally have larger resource limitations than intended. (or at least there is some way to ignore such services)

ref: https://github.com/line/armeria/blob/067578f161d43cc981b3b5c97f9fd385131c77cd/core/src/main/java/com/linecorp/armeria/server/ServiceConfigBuilder.java#L324-L351

Reply:
Good point. From an implementation perspective, it doesn't seem to be a problem that `GraphqlWebSocketService` inherits `WebSocketService`. The genuine problem is that the configuration is hard-coded with the `WebSocketService.class`.

In order not to set an unexpected configuration to `Graphql(WebSocket)Service`, we need to change the code to apply the default WebSocket settings only for `DefaultWebSocketService`.

Had a chat with other folks and we decided to provide dynamic options for a service. `RequestOptions` is going to be added to `HttpService`.
```java
interface HttpService {
   default RequestOptions requestOptions(RoutingContext routingContext) {
     return RequestOptions.of();
   }
}

interface WebSocketService {
   default RequestOptions requestOptions(RoutingContext routingContext) {
     return new WebRequestOptions(...);
   }
}

class GraphqlWebSocketService {
   default RequestOptions requestOptions(RoutingContext routingContext) {
      if (isWebSocket(routingContext) { 
        return new WebRequestOptions(...);
      } else {
        return RequestOptions.of();
      }
   }
}
```